### PR TITLE
ng_net: header building facilities

### PR DESCRIPTION
**Pull request by authmillenon**
This was taken out of #2553 

The idea is that an upper-layer can build a header for the layer it wants to send data over without knowledge about the actual structure of the header. I also added a convinience function of similar format to `ng_netif_hdr`.

~~Depends on #2706~~ (merged)

### CODE REVIEW

File: sys/include/net/ng_netif/hdr.h [hunk 1]
```diff
@@ -134,6 +136,46 @@ static inline void ng_netif_hdr_set_dst_addr(ng_netif_hdr_t *hdr, uint8_t *addr,
     memcpy(((uint8_t *)(hdr + 1)) + hdr->src_l2addr_len, addr, addr_len);
 }
 
+/**
+ * @brief   Builds a generic network interface header for sending and
```
Comment (OlegHahm) [inline sys/include/net/ng_netif/hdr.h]: s/network interface/network layer/` ? But what does "generic" mean in this context?

Comment (authmillenon) [inline sys/include/net/ng_netif/hdr.h]: No this is correct. This module (`ng_netif_hdr`) defines the "generic network interface header". This name was introduced after @OlegHahm was not happy with the name "generic link layer header" ;-)

Comment (OlegHahm) [inline sys/include/net/ng_netif/hdr.h]: I'm confused. I thought this was meant to implement (in a IMO weird way) IPv6 upper layer checksum calculation. Why do one need to build a generic network interface header?

Comment (authmillenon) [inline sys/include/net/ng_netif/hdr.h]: No that's in #2553. This just offers `netif_hdr_init()` and `pktbuf_add()` in one go.

File: sys/include/net/ng_netif/hdr.h [hunk 2]
```diff
@@ -134,6 +136,46 @@ static inline void ng_netif_hdr_set_dst_addr(ng_netif_hdr_t *hdr, uint8_t *addr,
...
+ *          adds it to the packet buffer.
+ *
+ * @author  Martine Lenders <mlenders@inf.fu-berlin.de>
```
Comment (OlegHahm) [inline sys/include/net/ng_netif/hdr.h]: Putting an author tag for this function on purpose? (I haven't even checked if this possible.)

Comment (authmillenon) [inline sys/include/net/ng_netif/hdr.h]: Yes. The rest of this header is by @haukepetersen.
And yes, of course it's possible.

**Comment by OlegHahm** [comment]
Sorry, I guess I asked this a 100 times before, but I'm confused again why an upper layer wants to build a lower layer's header?

**Comment by authmillenon** [comment]
Use-case: IPv6 wants to send packet.
* IPv6 checks next hop
* IPv6 searches link-layer address for next hop
* wants to tell link layer were to send it (and maybe additional information): uses link-layer header.

**Comment by authmillenon** [comment]
Why not just the address? Because in the rest of the model we use the header for exactly the same reason as I sketched out previously: The upper-layer might want to set additional information (hop-limit, traffic class, etc. you name it).

**Comment by OlegHahm** [comment]
> The upper-layer might want to set additional information (hop-limit, traffic class, etc. you name it).

But this would require an additional function, right? Why can I do this not just with a set_opt command?

**Comment by authmillenon** [comment]
Because this requires additional context switches. And no it would not need additional functions. Once you've build the header, you know the type of the header anyways, so if you need to, you can just change the fields after the creation. This functions are primarily for service functions for layers that DON'T need to know the build-up of a lower-layer header.

**Comment by authmillenon** [comment]
Rebased to current master

File: sys/include/net/ng_netreg.h [hunk 3]
```diff
@@ -126,6 +127,26 @@ int ng_netreg_num(ng_nettype_t type, uint32_t demux_ctx);
  */
 ng_netreg_entry_t *ng_netreg_getnext(ng_netreg_entry_t *entry);
 
+/**
+ * @brief   Builds a header for sending for a protocol in the registry and
```
Comment (Lotterleben) [inline sys/include/net/ng_netreg.h]: nitpick: I find this sentence a bit hard to parse. Maybe the “for sending” could be left out?

File: sys/net/crosslayer/ng_netreg/ng_netreg.c [hunk 4]
```diff
@@ -101,4 +102,30 @@ ng_netreg_entry_t *ng_netreg_getnext(ng_netreg_entry_t *entry)
...
 
+ng_pktsnip_t *ng_netreg_hdr_build(ng_nettype_t type, ng_pktsnip_t *payload,
+                                  uint8_t *src, uint8_t src_len,
+                                  uint8_t *dest, uint8_t dest_len)
+{
+    switch (type) {
+#ifdef MODULE_NG_IPV6
+
+        case NG_NETTYPE_IPV6:
+            return ng_ipv6_hdr_build(payload, src, src_len, dest, dest_len);
```
Comment (Lotterleben) [inline sys/net/crosslayer/ng_netreg/ng_netreg.c]: This seems to rely on ``sys/include/net/ng_ipv6/hdr.h``, which isn't in the master yet, but https://github.com/RIOT-OS/RIOT/pull/2454, which includes this file, relies on this PR... I'm sure that's according to your plans, but I'm a bit confused.

File: sys/include/net/ng_netreg.h (hunk 3 above)
Comment (authmillenon) [inline sys/include/net/ng_netreg.h]: No, since a receiving packet would store the payload differently.

File: sys/net/crosslayer/ng_netreg/ng_netreg.c (hunk 4 above)
Comment (authmillenon) [inline sys/net/crosslayer/ng_netreg/ng_netreg.c]: Since it is guarded by the `#ifdef MODULE_NG_IPV6` guard this line will only be compiled if the new IPv6 module is too.

File: sys/include/net/ng_netreg.h (hunk 3 above)
Comment (Lotterleben) [inline sys/include/net/ng_netreg.h]: Oh. Then maybe we could split it up into two sentences or something?

File: sys/net/crosslayer/ng_netreg/ng_netreg.c (hunk 4 above)
Comment (Lotterleben) [inline sys/net/crosslayer/ng_netreg/ng_netreg.c]: True.

File: sys/include/net/ng_netreg.h (hunk 3 above)
Comment (authmillenon) [inline sys/include/net/ng_netreg.h]: "Builds a header for sending and adds it to the packet buffer" would also work, I think. That the protocol is defined through `type` is apparent through the type parameter I think.

Comment (authmillenon) [inline sys/include/net/ng_netreg.h]: Done

**Comment by authmillenon** [comment]
Simplified doc a little.

**Comment by authmillenon** [comment]
Needed to add #2706 as a depenency since Travis is taking forever -.-

**Comment by Lotterleben** [comment]
Awww come on Travis!

**Comment by Kijewski** [comment]
The last successful Travis run is two and a half days old. I think we've killed that poor fellow.

**Comment by saurabh984** [comment]
I'll sponsor a dedicated server if anyone fancies setting up Janky (or Jenkins) on it! Travis is a nighmare

**Comment by authmillenon** [comment]
@saurabh984 we're (or rather @phiros is) working on a solution. We had [Jenkins](https://ci.riot-os.org/jenkins/) in the past. This was also no solution, since someone(TM) needs to manage that machine and nobody got time for that.

**Comment by authmillenon** [comment]
It's probably gonna be build bot.

**Comment by authmillenon** [comment]
But thanks for the offer

**Comment by authmillenon** [comment]
Fixed thingie

File: sys/include/net/ng_netif/hdr.h (hunk 2 above)
Comment (haukepetersen) [inline sys/include/net/ng_netif/hdr.h]: I don't really care, but you could just stick to our usual convention (it is a convention, right?), to just add your name to the file header once you changed stuff in the file. Or any particular reason in this case to do it differently?

**Comment by authmillenon** [comment]
Moved author tag.

**Comment by authmillenon** [comment]
Apart from the squash and PR dependency warning and an aborted (and now restarted) build in the msp430 group travis seems to be fine. Squash?

**Comment by Lotterleben** [comment]
Yes please!

**Comment by authmillenon** [comment]
Done

**Comment by Lotterleben** [comment]
:+1: Okay... Let's do this! ACK & Go.

**Comment by Lotterleben** [comment]
Correction: ACK, wait for Travis & go. I'll keep this tab open.

**Comment by authmillenon** [comment]
Ahhh sorry… I needed to change something for constistency: `s/dest/dst/`

**Comment by Lotterleben** [comment]
:D good catch! Tab is still open, will merge as soon as Travis is happy.

**Comment by OlegHahm** [comment]
I would like to put this topic on the list for the next NSTF meeting. I'm not nacking this, but am still not convinced of this approach.

**Comment by authmillenon** [comment]
@OlegHahm can you point me to the problem with this approach?

**Comment by OlegHahm** [comment]
No, I just still don't like upper layers building lower layers' headers. But it's so far more like a gut feeling.
